### ISSUE TITLE
Order of srcs matters to c2hs

### ISSUE THREAD

**Issue by mboes** [tweag/rules_haskell#344]
Consider `//tests/chs`. If you apply the following patch to the `BUILD` file, the test will fail!

```
diff --git i/tests/c2hs/BUILD w/tests/c2hs/BUILD
index 4e6e540..e26655d 100644
--- i/tests/c2hs/BUILD
+++ w/tests/c2hs/BUILD
@@ -16,8 +16,8 @@ haskell_cc_import(
 haskell_library(
   name = "c2hs",
   srcs = [
-    "Foo.chs",
     "Bar.chs",
+    "Foo.chs",
   ],
   deps = [
     ":zlib",

```

It fails like this:

```
$ bazel test //tests/c2hs
INFO: Analysed target //tests/c2hs:c2hs (1 packages loaded).
INFO: Found 1 target and 0 test targets...
ERROR: /home/mboes/src/tweag/rules_haskell/tests/c2hs/BUILD:16:1: HaskellC2Hs tests/c2hs/chs-c2hs/Bar.hs failed (Exit 1)
c2hs: Foo.chi not found in:
bazel-out/k8-fastbuild/bin/tests/c2hs/chs-c2hs
.

Target //tests/c2hs:c2hs failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.535s, Critical Path: 0.02s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

This is probably an upstream issue, but we should either document it or better yet, work around it.

**Comment by mrkkrp** [comment]
This is not a bug imo, it's just how c2hs works. `chs` files rely on `chi` interface files (or something like that) created as a side effect of pre-processing of other such files. There is no way to "work-around" this, I guess it's just a design decision of c2hs. I propose we close this as wonfix, actually.

**Comment by mboes** [comment]
But the end result is that order in `srcs` matters, when it shouldn't and the Bazel tooling is not designed to respect that. In particular, `buildifier` blithely reorders `srcs` entries in alphabetical order. So if `srcs` is order sensitive then people can't use `buildifier`.

A solution is to introduce `haskell_c2hs_library`. That way if there are dependencies between `.chs` files,  the user has to specify that themselves. It solves the order-dependence we're seeing here.

**Comment by mboes** [comment]
Fixed by #351.
